### PR TITLE
[CPU] Add hierarchical all-reduce with RDMA for multi-node tensor parallelism

### PR DESCRIPTION
**Pull request by wenlujon**
## Purpose

This PR adds an opt-in **Hierarchical All-Reduce (HAR)** path to the CPU backend to make **multi-node tensor parallelism** practical on Arm Grace (and other CPU) clusters.

Today the CPU backend reduces activations with a flat gloo/TCP ring all-reduce. On a 2-node setup this crosses the slow inter-node link on every collective (~96 collectives/token for Qwen2.5-14B), so decode becomes network-bound and throughput collapses (5.47 tok/s at TP=4). A flat ring is also topology-blind and, at TP=4, fails to make cross-node TP usable.

HAR replaces the flat ring with a two-level collective:
- **Intra-node:** shared-memory (SHM) reduce — no syscalls, no NIC.
- **Inter-node:** a single **RDMA Write-with-Immediate** hop over libibverbs (kernel-bypass, zero-copy), with each rank owning its own queue pair (4-way direct exchange, no leader relay).

The PR also includes two related changes:
- **Topology-aware logits all-reduce:** the per-token vocab-wide logits gather (the last collective still on gloo/TCP) is re-routed onto the HAR/RDMA path when ranks span nodes; single-node deployments keep the original fast SHM reduce.
- **Dynamic RDMA buffer sizing:** the registered RDMA region is right-sized to the model (≈148 MB vs a 256 MB static default for Qwen2.5-14B) to stay resident in the NIC's MTT cache and avoid an address-translation thrashing regression.

The entire RDMA path is gated behind a CMake build option **`VLLM_CPU_RDMA_HAR` (default `OFF`)**. With the flag off, all RDMA symbols are `#ifdef`-compiled out and the build is byte-for-byte the original SHM-only CPU path — **zero impact on stock CPU builds** and portable to nodes without InfiniBand.

### Files changed
| File | Change |
| --- | --- |
| `CMakeLists.txt`, `cmake/cpu_extension.cmake` | Add `VLLM_CPU_RDMA_HAR` build guard; conditionally link `libibverbs` |
| `csrc/cpu/shm.cpp` | RDMA HAR implementation (exchange → SHM barrier → local reduce), dynamic buffer sizing |
| `csrc/cpu/torch_bindings.cpp` | Bindings for the HAR all-reduce ops |
| `vllm/distributed/device_communicators/cpu_communicator.py` | HAR collective wiring, cross-node detection |
| `vllm/distributed/device_communicators/shm_broadcast.py` | SHM broadcast adjustments |
| `vllm/distributed/parallel_state.py` | Pass `local_world_size` for same-node detection |
| `vllm/model_executor/layers/logits_processor.py` | Topology-routed logits all-reduce |
| `vllm/v1/executor/multiproc_executor.py` | Forward `local_world_size` to workers |

## Test Plan

**Environment**
- 2 × NVIDIA Grace nodes, Arm Neoverse-V2 (aarch64), 144 cores (2×72, 2 NUMA), 942 GiB RAM/node
- NIC: Mellanox ConnectX (MT4125), RDMA via libibverbs; gloo device `mlx5_1`
- OS Ubuntu 24.04.4 (kernel 6.8.0), GCC 14.3, Arm Compute Library, Python 3.12, PyTorch 2.11
- Model: `Qwen/Qwen2.5-14B-Instruct` (bf16); workload: sharegpt (10 prompts)

**Cases**
1. Build with `-DVLLM_CPU_RDMA_HAR=OFF` (default) — confirm single-node TP=1 / TP=2 SHM baseline is unchanged (no regression).
2. Build with `-DVLLM_CPU_RDMA_HAR=ON` — run TP=2 and TP=4 across the two nodes, comparing flat TCP, pure RDMA, and HAR.
3. Accuracy: exact-match over the benchmark set + GSM8K.

## Test Result

**Decode output throughput — Qwen2.5-14B, 10-prompt sharegpt (best run per config)**

| Config | Transport | tok/s |
| --- | --- | ---: |
| TP=1, single node | — | 23.68 |
| TP=2, 1 node | SHM | 32.49 |
| TP=2, 2 nodes | TCP gloo (baseline) | 11.74 |
| TP=2, 2 nodes | RDMA | 18.68 |
| TP=2, 2 nodes | **HAR** | **34.40** |
| TP=4, 2 nodes | TCP gloo (baseline) | 5.47 |
| TP=4, 2 nodes | **HAR** | **40.80** |

**TP=4 optimization ladder (median output tok/s)**

| Stage | tok/s | Δ |
| --- | ---: | --- |
| Baseline (flat TCP gloo) | 5.47 | — |
| + HAR RDMA all-reduce | 25.4 | 4.6× |
| + Topology-routed logits all-reduce | 36.2 | +42% |
| + `GOMP_SPINCOUNT=100M` (runtime tuning) | 40.8 | +13% |

- **7.45× end-to-end** over the TCP baseline at TP=4 (5.47 → 40.8 tok/s); TP=4 HAR exceeds the single-node SHM ceiling (32.5 tok/s).
- **Accuracy preserved:** 50/50 exact-match vs the baseline (no quality loss); GSM8K 0.96.
- **Feature OFF:** single-node TP=1 and TP=2 SHM throughput identical to upstream — no regression.

> Note: `GOMP_SPINCOUNT=100M` is a runtime OpenMP tuning (keeps pinned cores hot between the many small decode kernels), not part of this code change; the code-attributable gain (HAR + logits routing) is 5.47 → 36.2 tok/s.

>Note: test based on changes against commit 8a5cf1ccd65e8ac7635c402c1ec0b08988bc26ca

## (Optional) Documentation Update

Building with the feature on:
```bash
VLLM_TARGET_DEVICE=cpu \
CMAKE_ARGS="-DVLLM_CPU_RDMA_HAR=ON" \
pip install -e . --no-build-isolation
```
Requires `libibverbs` and an RDMA-capable fabric (InfiniBand/RoCE). When `VLLM_CPU_RDMA_HAR=OFF` (default) the backend is unchanged and has no IB dependency.

### CODE REVIEW

**Comment by wenlujon** [comment]
@njhill @tlrmchlsmth @LucasWilkinson @Harry-Chen Hi Folks, could you please review the PR?
